app/peerinfo/peerinfo.go

```diff
@@ -0,0 +1,165 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package peerinfo
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/obolnetwork/charon/app/log"
+	pbv1 "github.com/obolnetwork/charon/app/peerinfo/peerinfopb/v1"
+	"github.com/obolnetwork/charon/app/z"
+	"github.com/obolnetwork/charon/p2p"
+)
+
+const period = time.Minute
+
+var protocolID protocol.ID = "/charon/peerinfo/1.0.0"
+
+type (
+	tickerProvider func() (<-chan time.Time, func())
+	nowFunc        func() time.Time
+)
+
+func New(tcpNode host.Host, peers []peer.ID, version string, lockHash []byte,
+	sendFunc p2p.SendReceiveFunc,
+) *PeerInfo {
+	tickerProvider := func() (<-chan time.Time, func()) {
+		ticker := time.NewTicker(period)
+		return ticker.C, ticker.Stop
+	}
+
+	return newInternal(tcpNode, peers, version, lockHash, sendFunc, p2p.RegisterHandler,
+		tickerProvider, time.Now)
+}
+
+func NewForT(_ *testing.T, tcpNode host.Host, peers []peer.ID, version string, lockHash []byte,
+	sendFunc p2p.SendReceiveFunc, registerHandler p2p.RegisterHandlerFunc,
+	tickerProvider tickerProvider, nowFunc nowFunc,
+) *PeerInfo {
+	return newInternal(tcpNode, peers, version, lockHash, sendFunc, registerHandler,
+		tickerProvider, nowFunc)
+}
+
+func newInternal(tcpNode host.Host, peers []peer.ID, version string, lockHash []byte,
```

Comment:
new internal what?
this function name seems vague

Reply:
also add godoc

Reply:
`New` and `NewForT` both call `newInternal`